Added image selection by variable along with other various improvements.

Hysrok opened the pull request:
> Added ability to designate which variable to use to create the image.
> Added min and max to input variable to control the scaling.
> Reworked the resolution variable so that it makes sense (currently 
represents the length of the longest side in pixels).
> Switched from using serializedSolution() to solution().
> Adjusted some of the color schemes.

Resolves issue #13780

<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

permcody committed:
Fix uninitialized SerializedSolution

closes #13763

Commit:
Added image selection by variable along with other various improvements.

> Added ability to designate which variable to use to create the image.
> Added min and max to input variable to control the scaling.
> Reworked the resolution variable so that it makes sense (currently 
represents the length of the longest side in pixels).
> Switched from using serializedSolution() to solution().
> Adjusted some of the color schemes.

Resolves issue #13780

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/371571/) on 504d90f wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/13782/site)

This comment will be updated on new commits.

permcody commented on framework/src/outputs/png/PNGOutput.C:
```diff
@@ -25,8 +26,12 @@ validParams<PNGOutput>()
+  params.addParam<unsigned int>("resolution", 25, "The resolution of the image.");
+  params.addParam<std::string>(
+      "variable", "None", "The name of the variable to use when creating the image");
+  params.addParam<Real>("max", -1000, "The maximum for the variable we want to use");
```
We have a way of testing whether a parameter was set by the user. If you use that you don't have to attempt to find a value that is "out of range" or invalid. I'd set this to one and the min to zero.

permcody commented on framework/src/outputs/png/PNGOutput.C:
```diff
@@ -58,34 +66,93 @@ void
+    mooseError("PNGOutput does not currently function in parallel.");
+
+  // By default _use_aux is false.
+  _use_aux = false;
```
Move this to the ctor initialization list. Never forget to initialize a POD type - it WILL bite you!

permcody commented on framework/src/outputs/png/PNGOutput.C:
```diff
@@ -58,34 +66,93 @@ void
+
+  // If a specific variable name isn't provided, skip this if statement and use the default
+  // varaible.
+  if (_variable != "None")
```
This should be a required parameter. I don't think it's a good default to grab the first variable in the system - that's not intuitive.

permcody commented on framework/src/outputs/png/PNGOutput.C:
```diff
@@ -58,34 +66,93 @@ void
+          _problem_ptr->getNonlinearSystem().getVariable(processor_id(), _variable).number();
+
+    else
+      mooseError("The given input variable '",
```
We use paramError() for this case (errors involving input file parameters) it gives contextual error information.

permcody commented on framework/src/outputs/png/PNGOutput.C:
```diff
@@ -25,8 +26,12 @@ validParams<PNGOutput>()
                         "Determination of whether the background will be transparent.");
-  params.addParam<unsigned int>("resolution", 1, "The resolution of the image.");
-  MooseEnum color("GRAY BRYW BCR RWB BR");
+  params.addParam<unsigned int>("resolution", 25, "The resolution of the image.");
```
addRangeCheckedParam here - make sure the resolution is non-negative. The zero case results in an FPE in your logic below. Negative is also bad.

permcody commented on test/tests/outputs/png/wedge.i:
```diff
@@ -17,6 +17,20 @@
+[]
+
+[AuxKernels]
+  [bnds]
```
I know this is picky by "bnds" is a bad name (even for phase-field it isn't great). Let's go with two_u or something like that.

Commit:
Fixes

> Applied changes found in PR request.
> Changed to use serialized solution again.

Commit:
Small adjustments.

> Changed variable for output in the adv_diff_reaction_test to be one 
that was actually in the input file.
> Adjusted the use of paramError to properly display the varailbe name 
when the error occurs.

Commit:
Fixed the getVariable call, which was not working when using multiple processors.

permcody commented on framework/src/outputs/png/PNGOutput.C:
```diff
@@ -58,34 +68,82 @@ void
+    _scaling_max = _max;
+
+  // If the min value wasn't specified in the input file, find it from the system.
+  if (_pars.isParamSetByUser("min"))
```
I believe you have the logic flipped here. If it *is* set by the user, you'll want to honor their value. Same for the check above.

Hysrok replied:
Oops!

permcody commented on framework/src/outputs/png/PNGOutput.C:
```diff
@@ -134,8 +192,9 @@ PNGOutput::setRGB(png_byte * rgb, Real selection)
     // RWB.  Change spectrum.
     case 3:
-      color_spectrum_max = 1013; // 511;
+      color_spectrum_max = 511;
```
Do you have any comments for these hardcoded values? They are odd.

Hysrok replied:
They are kinda odd.  I'll add explanatory comments.

permcody commented on framework/src/outputs/png/PNGOutput.C:
```diff
@@ -259,12 +306,34 @@ void
-  Point maxPoint = _box.max();
-  Point minPoint = _box.min();
+  Point max_point = _box.max();
+  Point min_point = _box.min();
```
Nice catch on the naming convention.

Hysrok replied:
:)

permcody commented on test/tests/outputs/png/wedge.i:
```diff
@@ -17,6 +17,20 @@
+[]
+
+[AuxKernels]
+  [tew_u]
```
Now that's funny. I'm not sure if you are just playing with the homophone here or not. I suggested "two_u" because this variable is double the solution of u (2*u) or two_u, "tew_u" wasn't quite what I was after - haha

Hysrok replied:
Yeah, I thought it would be fun to have that homophone in there for one commit :). I'll switch it out now.

permcody reviewed (changes requested):
One minor fix with the logic.

Commit:
Reworked color spectrum.

> Changed the comment descriptions of color_spectrum_max to aid 
understandibility.
> Added new variable to the colors function to aid in readability.

Commit:
Took out a redundant test.

Commit:
Fixed possible access of number_destination_colors before initialization
